### PR TITLE
Add SessionReceiver/SessionProcessor types

### PR DESCRIPTION
**Pull request by JoshLove-msft**
- Add separate types for SessionReceiver/SessionProcessor
- Don't expose receiver in processor event args
- Add DeadLetterErrorReason/Description properties to ServiceBusReceivedMessage
- Misc. clean up
- Add more tests

Fixes:
https://github.com/Azure/azure-sdk-for-net/issues/10724
https://github.com/Azure/azure-sdk-for-net/issues/10723
https://github.com/Azure/azure-sdk-for-net/issues/10699
https://github.com/Azure/azure-sdk-for-net/issues/10797

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs [hunk 1]
```diff
@@ -385,6 +423,29 @@ protected ServiceBusProcessor()
...
+                throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, Resources1.CannotStartEventProcessorWithoutHandler, nameof(ProcessErrorAsync)));
+            }
+        }
+
+        private void ValidateMessageHandler()
+        {
+            if (IsSessionProcessor)
+            {
+                if (_processSessionMessage == null)
+                {
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs]: nit: We can merge these if conditions.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        /// <summary>
+        /// Indicates whether or not this <see cref="ServiceBusProcessor"/> has been closed.
+        /// </summary>
+        ///
+        /// <value>
+        /// <c>true</c> if the client is closed; otherwise, <c>false</c>.
+        /// </value>
+        ///
+        public bool IsProcessing => _innerProcessor.IsProcessing;
+
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs]: `IsProcessing` would be true if the client is closed? It sounds weird. Did you mean something like `isClosed`/`isProcessorClosed`?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusReceivedMessage.cs [hunk 3]
```diff
@@ -256,6 +256,35 @@ public DateTimeOffset ExpiresAt
...
+            get
+            {
+                if (Properties.TryGetValue(DeadLetterErrorDescriptionHeader, out object description))
+                {
+                    return description as string;
+                }
+                return null;
+            }
+        }
         /// <summary>
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusReceivedMessage.cs]: nit: skip one line.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Client/ServiceBusClient.cs [hunk 4]
```diff
@@ -213,19 +216,54 @@ public ServiceBusClient(string fullyQualifiedNamespace, TokenCredential credenti
...
+        /// <param name="options"></param>
+        /// <returns></returns>
+        public ServiceBusReceiver GetDeadLetterReceiver(
+            string queueName,
+            ServiceBusReceiverOptions options = default) =>
+            new ServiceBusReceiver(
+                connection: Connection,
+                entityPath: EntityNameFormatter.FormatDeadLetterPath(queueName),
+                isSessionEntity: false,
+                options: options);
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Client/ServiceBusClient.cs]: Nope, the dead letter receiver is the same for session or non-session.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ProcessMessageEventArgs.cs [hunk 5]
```diff
@@ -22,23 +24,116 @@ public class ProcessMessageEventArgs : EventArgs
...
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="message"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task CompleteAsync(
+            ServiceBusReceivedMessage message,
+            CancellationToken cancellationToken = default) =>
+            await _receiver.CompleteAsync(
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ProcessMessageEventArgs.cs]: Correct, I figured this was not going to be a common use case for the processor.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs (hunk 2 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs]: Oh the docs need updating.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusSessionReceiver.cs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        /// </remarks>
+        ///
+        /// <param name="lockToken">The lockToken of the <see cref="ServiceBusReceivedMessage"/> to renew the lock for.</param>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        public override Task<DateTimeOffset> RenewMessageLockAsync(
+            string lockToken,
+            CancellationToken cancellationToken = default)
+        {
+            throw new InvalidOperationException(Resources.CannotLockMessageOnSessionEntity);
+        }
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusSessionReceiver.cs]: Removed from here and now handling in ServiceBusReceiver.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs (hunk 1 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs]: We wouldn't want to go into the else block if it is a session processor that has _processSessionMessage populated. We should only go to the else block if not a session processor.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs [hunk 7]
```diff
@@ -368,7 +368,7 @@ public virtual async Task<ServiceBusReceivedMessage> PeekAsync(CancellationToken
         /// </remarks>
         ///
         /// <returns>A task to be resolved on when the operation has completed.</returns>
-        public virtual async Task CompleteAsync(
+        internal virtual async Task CompleteAsync(
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: Just curious, why did we change this to internal?

Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs]: Ramya mentioned we should leave the batch settlement out of Preview 1.

**Review by ShivangiReja — APPROVED** [review]

**Pull request merged**